warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java

```diff
@@ -63,143 +67,91 @@ public AllTermsIndexedVisitor(ShardQueryConfiguration config, MetadataHelper hel
-                    newNode.jjtAddChild(newChild, newIndex);
-                    newIndex++;
-                }
-            }
-        }
-        
-        switch (newNode.jjtGetNumChildren()) {
-            case 0:
-                return null;
-            case 1:
-                JexlNode child = newNode.jjtGetChild(0);
-                JexlNodes.promote(newNode, child);
-                return child;
-            default:
-                return newNode;
-        }
+        return visitJunctionNode(node, data);
     }
     
     @Override
     public Object visit(ASTAndNode node, Object data) {
-        ASTAndNode newNode = new ASTAndNode(ParserTreeConstants.JJTANDNODE);
-        newNode.image = node.image;
+        return visitJunctionNode(node, data);
+    }
+    
+    private Object visitJunctionNode(JexlNode node, Object data) {
+        JexlNode copy = getCopyWithVisitedChildren(node, data);
         
-        int newIndex = 0;
-        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
-            Node newChild = (Node) node.jjtGetChild(i).jjtAccept(this, data);
-            
-            if (newChild != null) {
-                // When we have an AND or OR
-                if ((newChild instanceof ASTOrNode || newChild instanceof ASTAndNode)) {
-                    // Only add that node if it actually has children
-                    if (0 < newChild.jjtGetNumChildren()) {
-                        newNode.jjtAddChild(newChild, newIndex);
-                        newIndex++;
-                    }
-                } else {
-                    // Otherwise, we want to add the child regardless
-                    newNode.jjtAddChild(newChild, newIndex);
-                    newIndex++;
-                }
-            }
-        }
-        switch (newNode.jjtGetNumChildren()) {
+        switch (copy.jjtGetNumChildren()) {
             case 0:
                 return null;
             case 1:
-                JexlNode child = newNode.jjtGetChild(0);
-                JexlNodes.promote(newNode, child);
+                JexlNode child = copy.jjtGetChild(0);
+                promote(copy, child);
                 return child;
             default:
-                return newNode;
+                return copy;
         }
     }
     
+    private JexlNode getCopyWithVisitedChildren(JexlNode node, Object data) {
+        // @formatter:off
+        JexlNode[] children = Arrays.stream(JexlNodes.children(node))
+                        .map(n -> (JexlNode) n.jjtAccept(this, data)) // Visit the node
+                        .filter(Objects::nonNull) // Filter out null nodes
+                        .filter(n -> (!JexlNodes.isAnd(n) && !JexlNodes.isOr(n)) || JexlNodes.isNotChildless(n)) // Filter out empty junction nodes
+                        .toArray(JexlNode[]::new);
+        // @formatter:on
+        JexlNode copy = JexlNodes.newInstanceOfType(node);
+        copy.image = node.image;
+        System.out.println("Copy: " + copy);
```

Comment:
Please remove System.out.println.  Use the logger if anything.

Reply:
Fixed.